exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/TypeInferenceUtils.java

```diff
@@ -0,0 +1,571 @@
+          .put(SqlTypeName.INTEGER, TypeProtos.MinorType.INT)
+          .put(SqlTypeName.BIGINT, TypeProtos.MinorType.BIGINT)
+          .put(SqlTypeName.FLOAT, TypeProtos.MinorType.FLOAT4)
+          .put(SqlTypeName.DOUBLE, TypeProtos.MinorType.FLOAT8)
+          .put(SqlTypeName.VARCHAR, TypeProtos.MinorType.VARCHAR)
+          .put(SqlTypeName.BOOLEAN, TypeProtos.MinorType.BIT)
+          .put(SqlTypeName.DATE, TypeProtos.MinorType.DATE)
+          .put(SqlTypeName.TIME, TypeProtos.MinorType.TIME)
+          .put(SqlTypeName.TIMESTAMP, TypeProtos.MinorType.TIMESTAMP)
+          .put(SqlTypeName.VARBINARY, TypeProtos.MinorType.VARBINARY)
+          .put(SqlTypeName.INTERVAL_YEAR_MONTH, TypeProtos.MinorType.INTERVALYEAR)
+          .put(SqlTypeName.INTERVAL_DAY_TIME, TypeProtos.MinorType.INTERVALDAY)
+          .put(SqlTypeName.CHAR, TypeProtos.MinorType.VARCHAR)
+
+          // The following types are not added due to a variety of reasons:
+          // (1) Disabling decimal type
+          //.put(SqlTypeName.DECIMAL, TypeProtos.MinorType.DECIMAL9)
+          //.put(SqlTypeName.DECIMAL, TypeProtos.MinorType.DECIMAL18)
+          //.put(SqlTypeName.DECIMAL, TypeProtos.MinorType.DECIMAL28SPARSE)
+          //.put(SqlTypeName.DECIMAL, TypeProtos.MinorType.DECIMAL38SPARSE)
+
+          // (2) These 2 types are defined in the Drill type system but have been turned off for now
+          // .put(SqlTypeName.TINYINT, TypeProtos.MinorType.TINYINT)
+          // .put(SqlTypeName.SMALLINT, TypeProtos.MinorType.SMALLINT)
+
+          // (3) Calcite types currently not supported by Drill, nor defined in the Drill type list:
+          //      - SYMBOL, MULTISET, DISTINCT, STRUCTURED, ROW, OTHER, CURSOR, COLUMN_LIST
+          // .put(SqlTypeName.MAP, TypeProtos.MinorType.MAP)
+          // .put(SqlTypeName.ARRAY, TypeProtos.MinorType.LIST)
+          .build();
+
+  /**
+   * Given a Drill's TypeProtos.MinorType, return a Calcite's corresponding SqlTypeName
+   */
+  public static SqlTypeName getCalciteTypeFromDrillType(final TypeProtos.MinorType type) {
+    return DRILL_TO_CALCITE_TYPE_MAPPING.get(type);
+  }
+
+  /**
+   * Given a Calcite's RelDataType, return a Drill's corresponding TypeProtos.MinorType
+   */
+  public static TypeProtos.MinorType getDrillTypeFromCalciteType(final RelDataType relDataType) {
+    final SqlTypeName sqlTypeName = relDataType.getSqlTypeName();
+    TypeProtos.MinorType minorType = CALCITE_TO_DRILL_MAPPING.get(sqlTypeName);
+    if(minorType == null) {
+      minorType = TypeProtos.MinorType.LATE;
+    }
+    return minorType;
+  }
+
+  /**
+   * Give the name and DrillFuncHolder list, return the inference mechanism.
+   */
+  public static SqlReturnTypeInference getDrillSqlReturnTypeInference(
+      final String name,
+      final List<DrillFuncHolder> functions) {
+    switch(name.toUpperCase()) {
+      case "DATE_PART":
+        return DrillDatePartSqlReturnTypeInference.INSTANCE;
```

Comment:
Right, most of them are singleton, except for the default one, which needs a DrillFuncHolder list to be associated with them.